modules/hal_wch/ch32_rcc.h

```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2024 Dhiru Kholia
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _CH32_RCC_H
+#define _CH32_RCC_H
+
+#ifdef CONFIG_SOC_CH32V003
+#include <ch32v003fun.h>
+#else
+#error "SoC not supported!"
+#endif
+
+#endif
```

Comment:
003fun is a single header that includes everything, but some people prefer to mimic a more traditional header structure.  I have seen this pattern frequently.  But, do not know if it is against or in line with Zephyr.

Reply:
This style was requested by another reviewer on this PR, so I'd be reluctant to revert it.

Reply:
It's difficult to tell for sure, but this was the comment where it was added. https://github.com/zephyrproject-rtos/zephyr/pull/73761#issuecomment-2389030027  I think it was part of a conversation with @fabiobaltieri but I don't recall for sure who the reviewer was who recommended it.

It can be difficult to know who to "trust more" on these matters.  If your recommendation takes precedence, then, I would expect @kholia would be up for switching it again?

Reply:
I believe it was @nandojve who initially requested this.

Reply:
> I don't recall about this.
> 
> Usually we prefer to use:
> 
> ```diff
> -#include <ch32v003fun.h>
> +#include <ch32_some_driver.h>
> ```

That only makes sense if the HAL splits stuff; otherwise it's just faking reality.

Reply:
Yes, and everything that is in mainline the hal is splitted by the SoC families. It doesn't matter if it is a hal API or registers files. But if we allow to use `#include <ch32v003fun.h>` how we deal with #64472 ? Will this vendor have a driver for each SoC? How we will guide people to add the next SoC or driver?

The problem that I see here is that the PR is breaking the "general rule" of having for each SoC a definition of their on HAL API implementation or register set. For me, still not clear how to deal with this and I did not look to see if registers names are align between PR's or even with vendor files. I believe that if we can respect the vendor registers names and the module/HAL deal with the files it could be easy to add another SoC and still reuse code.

I'm not sure if this could require an update on the guidelines to provide better guidance on this details.

I let to you @gmarull decide what fits better here.

Reply:
ch32v003fun.h is for all CH32Vxxx and CH32Xxxx series chips, it differentiates by the `#define`'d architecture.

My guess is that it would be less of a diff to add many new arch's if you didn't need a vendored set of headers per-SoC.  As they would all just be stubs.

Reply:
Hi @cnlohr ,

If that is the case:

- Why not just add the necessary registers/defines direct inside the driver as @xingrz is doing?
- Why we should keep a centralized place? What we gain with that? We already have this in the past when drivers include soc.h and we already move forward.
- Why we should maintain a HAL with only one file https://github.com/zephyrproject-rtos/hal_wch/blob/main/ch32v003fun/ch32v003fun.h of 13145 lines?

I vote to have:
 - driver self contained:
   - Option 1: All definitions inside the driver.
   - Option 2: Include 1 well defined header from HAL with the definitions to the peripheral.
 - SoC self contained. All definitions inside the family folder inside the soc directory (preferable). The soc.h can be imported from HAL too but I'm trying to eliminate completely the HAL for this family since this seems to make more sense at moment (no strong opinion).

This way it is possible to have each driver self contained and can be reusable, no SoC dependency. The idea is split the 13145 lines of the ch32v003fun.h definitions in multiple files per peripheral and per SoC family. Things like below c snip should be forbidding. In Zephyr we select specifics using the SoC family/series Kconfigs and the below seems to be a remaining of baremetal. Since there is a HAL it is possible to use direct the CONFIG_SOC_CH32V003 inside of it or pass by c compiler as -DCH32V003=1

```c
#ifndef _FUNCONFIG_H
#define _FUNCONFIG_H

#ifdef CONFIG_SOC_CH32V003
#define CH32V003 1
#endif

#endif
```

If you need the HAL content to be used externally from Zephyr you still can include the spitted files.

I suggest the following hal structure if you prefer split on files:

```c
include/wch/soc-family/soc-family.h
include/wch/drivers/ch32-prefix_driver-type.h
```

You should structure the soc Kconfigs in a way that allow you to use the Family/Series if you prefer to have only one soc-family.h header.

CC: @gmarull @cfriedt @xingrz @soburi

Reply:
There really is no need to add a hal if it can be easily avoided.

Adding the necessary defines inside of each driver can be much easier.

Reply:
I believe I understand your comment @nandojve , but two parts - 
1. You can treat ch32v003fun.h as vendored.  
And - 
2. It is common for the split you are describing in many older chip libraries, such as ST.  But much of the C ecosystem is moving towards single-file-header-libraries and this experience with more of an amalgam has been very positive. The discourse around it has also been positive on the discord and github. Allowing inclusion of a whole platform with nothing more than a compiler and 4 files, has a variety of benefits... With the only drawback being that it's not immediately clear what to read to get a handle on everything.

My feeling is that zephyr would benefit from not splitting it out into a plethora of HAL files, and instead just including the needed config file and the ch32v003fun.h. And, if I am interpreting @cfriedt and @gmarull 's comments correctly, they are more aligned with the approach that requires many fewer files.  

But, again, I don't know how zephyr determine architecture direction - and I think either solution is completely acceptable for developers, and am not sure why that would hold up this PR.

Reply:
Thanks folks.

I will reduce the number of these `modules/hal_wch/ch32_*.h` files. Perhaps, a single `modules/hal_wch/ch32_wrapper.h` would be enough (naming is hard and suggestions are welcome here).

Reply:
I don't think you need to do any big change here. Just create a series agnostic header that wraps all series, something like:

```c
/* ch32fun.h */
...
#if defined(CONFIG_SOC_SERIES_X)
#include <ch32v00xfun.h>
#elif defined(CONFIG_SOC_SERIES_Y)
#include <ch32v00yfun.h>
...
#endif
```

So that drivers just include `ch32fun.h` when they need to access any register (note: names may be adjusted).

While it would be nice to have a generic way to access registers (e.g. something like svd2rust or chiptool), I don't think it's the scope of this PR. Most vendors don't even bother writing clean, native drivers, and we're left with HAL hackery.